Typescript 1st challenge solved

gcoayla opened the pull request.

gcoayla committed:
Challenge completed

gcoayla committed:
Added comments to the functions

donovanhiland commented on src/app.ts:
```diff
@@ -13,14 +14,73 @@ interface TemperatureSummary {
 export function processReadings(readings: TemperatureReading[]): void {
-  // add here your code
+  for (let read of readings) {
+    let exist = true
+    if (read.city in sumarys) {
+      if (!(read.time.toDateString() in sumarys[read.city])) {
+        exist = false
+      }
+    } else {
+      exist = false
+    }
```
I'm betting you can do all this in a single line

donovanhiland commented on src/app.ts:
```diff
@@ -13,14 +14,73 @@ interface TemperatureSummary {
+          last: cityLast.temperature,
+          high: cityHigh.temperature,
+          low: cityLow.temperature,
+          average: cityAvr,
```
I'm going to update the challenge wording, but this average is the average for the day. Lets also add another `average` field on to the city itself with the all time average temp

donovanhiland commented on src/app.ts:
```diff
@@ -13,14 +14,73 @@ interface TemperatureSummary {
   city: string,
 ): TemperatureSummary | null {
-  //add here your code
+  if (city in sumarys) {
+    if (date.toDateString() in sumarys[city]) {
+      return sumarys[city][date.toDateString()]
+    }
+  }
```
Sometimes I like if statements for readability, but other times like here I'm leaning towards just putting the conditions on a single line

gcoayla committed:
Changed if statements and variable name cityAvr to cityDayAvr

donovanhiland commented on src/app.ts:
```diff
@@ -13,14 +14,61 @@ interface TemperatureSummary {
 export function processReadings(readings: TemperatureReading[]): void {
-  // add here your code
+  for (let read of readings) {
+    let exist = read.city in sumarys ? (read.time.toDateString() in sumarys[read.city]) ? true : false : false
```
I would skip the ternary here and just use a logic gate

```suggestion
    let exist = read.city in sumarys && (read.time.toDateString() in sumarys[read.city])
```

donovanhiland replied:
Also since you're using this same logic elsewhere I would probably pull it out into a function

donovanhiland commented on src/app.ts:
```diff
@@ -13,14 +14,61 @@ interface TemperatureSummary {
+      )
+      let cityFirst = cityReadings[0]
+      let cityLast = cityReadings[cityReadings.length - 1]
+      let cityHigh = cityReadings.reduce((a, b) =>
+        a.temperature > b.temperature ? a : b,
+      )
+      let cityLow = cityReadings.reduce((a, b) =>
+        a.temperature < b.temperature ? a : b,
+      )
```
The high and the low firstly should be for the day. What is the high for today and hat is the low for today. You can also add in a high and low for the city for all the readings. Either way I don't think you need to use reduce. Reduce is just adding to the complexity of the function by adding extra loops to each loop. `Math.min` and `Math.max` between 2 numbers should be good enough

gcoayla committed:
Move the if exist condition to a function and changed variable names

gcoayla committed:
Added 'historic' record for each city

TemperatureSummary now has two optional properties that helps to store the Historic average.

Each city now have an 'historic' key in the dictionary that it's also a TemperatureSummary object, and stores the historical low, high and average temperature of the city.

The average historical temperature its calculated by the division between the accumAvr and regCount.
